File: Témata/02 NLRM s nezávislou normální-gama apriorní hustotou (Gibbs)/Matlab/commute_testy2.m
clear
close all
clc

%% Pridani cesty k podpurnym funkcim v adresari Support
% adresar pro vlastni funkce (pokud nejsou ve stejnem adresari jako skript)
 addpath('E:\MUNI\projekt_bayes\Témata\Support\Matlab')

%% Nacteni dat
%pozorovani za 231 pracovnich dni v roce 2006)

% time      ... cas cesty
% depart    ... cas odchodu (v minutach od 6:30)
% reds      ... pocet cerevenych svetel na semaforu
% trains    ... pocet vlaku, ktere musi nechat projek na Murrumbeena prejezdu


load data_commute.mat

%% Nastaveni apriornich hyperparametru a Gibbsova vzorkovace
% Apriornimi hyperparametry
% p(beta)~N(beta_0, V_0)
% p(h)~G(h_0,nu_0)
 beta_0 = [30
           1
           3
           5];
 V_0 = diag([7.5^2;0.25^2;1^2;2^2]);
 nu_0 = 40;
 s2_0 = 10^2;
 h_0 = 1/s2_0;
 
% Definice modelu
 y = time;
 X = [ones(size(y)) depart reds trains];

% Nastaveni Gibbsova vzorkovace
 S = 50000+1;   %celkovy pocet generovanych vzorku + pocatecni hodnota
 S_0 = 30000+1; %pocet vyhozenych vzorku
 S_1 = S-S_0;   %pocet ponechanych vzorku

 beta = zeros(length(beta_0),S);    %vzorky pro beta
 h = zeros(1,S);                    %vzorky pro h
  
% nastaveni pocatecnich hodnot
 beta(:,1) = beta_0;
 h(1,1) = h_0;
 
%% Dalsi deklarace vektoru, napr. pro Savage-Dickey pomer hustot
% Jmenovatel pro SD pomer hustot - priklad d)
SDd_nom = zeros(1,S);
% Jmenovatel pro SD pomer hustot - priklad e)
SDe_nom = zeros(1,S);

count_post_restrict = 0; %pocitadlo nesplneni podminek apriornich
                         %restrikci na parametry

%% Gibbsuv vzorkovac
for s=2:S
 %1. blok Gibbsova vzorkovace
 %podminena hustota p(beta|h,y)~N(beta_1,V_1)
  V_1 = inv(inv(V_0)+h(1,s-1)*(X'*X)); %(4.4) dle Koop (2003)
  beta_1 = V_1*(inv(V_0)*beta_0+h(1,s-1)*(X'*y)); %(4.5) dle Koop (2003)
  
  check_restrict = 0; %nastaveni promenne overujici splneni
                      %podminky apriornich restrikci 
                      %na hodnoty parametru (0 = nesplneno)
  while check_restrict==0 %cyklus se opakuje dokud neni splnena
                          %apriorni restrikce na parametry
    beta(:,s) = beta_1+norm_rnd(V_1); %(4.7) dle Koop (2003)
    count_post_restrict = count_post_restrict+1;
    if prod(beta(:,s)>=0)==1 %soucin vyhodnoceni podminek kladnosti
        %na parametry musi byt roven 1 (podminky lze menit dle
        %potreby dle apriornich omezeni - nutno je v ramci
        %vypoctu S-D pomeru hustot zohlednit v apriornich hustotach
       check_restrict = 1; 
       count_post_restrict = count_post_restrict-1;
        %korekce pocitadla
    end
  end
 %2. blok Gibbsova vzorkovace
 %podminena hustota p(h|beta,y)~G(h_1,nu_1)
  nu_1 = length(y)+nu_0;            %(4.9)
  h_1 = (1/nu_1*((y-X*beta(:,s))'*(y-X*beta(:,s))+nu_0*1/h_0))^-1; %(4.10)
 
  h(1,s) = gamm_rnd_Koop(h_1,nu_1,1); %(4.8)
  
 %Cast pro vypocet citatelu SD pomer hustot
 %d) beta_depart = 0 (druhy prvek vektoru beta
 % p(b|h,y) = N(beta_1,V_1)
  V_1 = inv(inv(V_0)+h(1,s)*(X'*X));
  beta_1 = V_1*(inv(V_0)*beta_0+h(1,s)*(X'*y));
  SDd_nom(s) = normpdf(0,beta_1(2),sqrt(V_1(2,2)));
 %e) beta_reds = beta_trains
 % p(R*b|h,y) = N(R*beta_1,R*V_1*R')
  R = [0 0 1 -1];
  r = 0;
  SDe_nom(s) = mvnpdf(r,R*beta_1,R*V_1*R');    
end

%% Posteriorni analyza
% vyhozeni prvnich S_0 vzorku
 beta(:,1:S_0) = [];
 h(:,1:S_0) = [];
 SDd_nom(:,1:S_0) = [];
 SDe_nom(:,1:S_0) = [];
 
% graficke zobrazeni konvergence
 k = 100;   %delka kroku
 figure
 for ii=1:length(beta_0)
    subplot(3,2,ii)
    plot(beta(ii,1:k:end));
    ylabel(['\beta_',num2str(ii)])
 end
  subplot(3,2,ii+1)
  plot(h(1:k:end))
  ylabel('h')

% Gewekova konvergencni diagnostika
 CD_beta = Geweke(beta');
 CD_h = Geweke(h');
 
%% Prezentace vysledku
%apriorni str. hodnoty a sm. odchylky
%beta_0, h_0 - apriorni stredni hodnoty
std_beta_0 = sqrt(diag(V_0)); %vektor apriornich sm. odchylek
std_h_0 = sqrt(2*h_0^2/nu_0); %apriorni sm. odchylka pro h

%posteriorni str. hodnoty a sm. odchylky
mean_beta_1 = mean(beta,2); %sloupcovy vektor radkovych prumeru
mean_h_1 = mean(h); %vystup = skalarni velicina
std_beta_1 = sqrt(mean(beta.^2,2)-mean_beta_1.^2);  
std_h_1 = sqrt(mean(h.^2)-mean_h_1.^2);

%Vystup na obrazovku
fprintf('Parametr  prior m.  prior std.  post m. post std. CD\n')
fprintf('====================================================\n')
for ii=1:length(beta_0)
fprintf('Beta %1.0f   %6.4f  %6.4f %6.4f %6.4f %6.4f\n',...
    ii,beta_0(ii),std_beta_0(ii),mean_beta_1(ii),...
    std_beta_1(ii),CD_beta.CD(ii)) 
end
fprintf('h        %6.4f  %6.4f %6.4f %6.4f %6.4f\n',...
    h_0,std_h_0,mean_h_1,std_h_1,CD_h.CD) 
    
%% test hypotezy ze beta_red (beta_2) >=2
%pro ilustraci: zobrazeni rozdeleni beta_2 pomoci histogramu
 figure
 histogram(beta(3,:))
 title('Posterní hustota parametru \beta_2')
 xlabel('\beta_2')
%pravdepodobnost beta_2>=2
 pc = sum(beta(3,:)>=2)/S_1;
 fprintf('\n Pravdepodobnost beta_red >= 2 a odpovidajici Bayesuv faktor\n')
 fprintf('\n Pravdepodobnost = %4.3f     Bayesuv faktor = %6.3f\n',...
     pc,pc/(1-pc))

%% d) Test hypotezy beta_depart = 0 (druhy prvek)radek vektoru/matice parametru beta)
%Savage-Dickey pomer hustot
%Vyhodnotime p(beta_depart=0|M2)
%apriorni hustota p(beta_depart)~N(beta_0(2),V_0(2,2))
%jmenovatel SD pomeru hustot
%vyuzijeme funkci statistickeho toolboxu normpdf 
%pro pdf normalniho rozdeleni
 SDd_denom = normpdf(0,beta_0(2),sqrt(V_0(2,2)));
%simulace integracni konstanty omezen apriorni hustoty
 S_sim = 50000;
 count_prior_restrict = 0; %pocitadlo nesplneni apriornich omezene
 for ii=1:S_sim
    check_restrict = 0;
    while check_restrict==0
     pom = beta_0(2)+norm_rnd(V_0(2,2));
     count_prior_restrict = count_prior_restrict+1;
     if pom>=0
        check_restrict=1;
        count_prior_restrict = count_prior_restrict-1;
     end
    end
 end
%integracni konstanta apriorni hustoty zohlednujici apriorni omezeni
int_prior_restrict = 1/(1-count_prior_restrict/...
    (S_sim+count_prior_restrict));
%korekce jmenovatele S-D pomeru hustot
 SDd_denom = int_prior_restrict*SDd_denom;
%vypocet citatele na zaklade vystupu Gibbsova vzorkovace (doplneneho)
 E_SDd_nom = mean(SDd_nom);
%integracni konstanta posteriorni hustoty zohlednujici
%apriorni omezeni
 E_SDd_nom = 1/(1-count_post_restrict/(S+count_post_restrict-1))...
     *E_SDd_nom;
%Bayesuv faktor
 SD_d = E_SDd_nom/SDd_denom;
fprintf('\n Bayesuv faktor pro beta_depart = 0 \n')
fprintf('BF = %6.4f\n',SD_d)

%% e) Test beta_red = beta_trains (treti a ctvrty prvek vektoru beta)
% Linearni omezeni R*beta = r
%matice omezeni (na cely vektor parametru beta
 R = [0 0 1 -1];
 r = 0;
%jmenovatel pro SD pomer hustot
%apriorni hustota p(beta)~N(beta_0,V_0)
%apriorni hustota p(R*beta)~N(R*beta_0,R*V_0*R')
%vyuziti funkce hustoty pro vicerozmerne normalni rozdeleni
%ze statistickeho toolboxu - mvnpdf
 SDe_denom = mvnpdf(r,R*beta_0,R*V_0*R');  

%simulace integracni konstanty omezen apriorni hustoty
 S_sim = 50000;
 count_prior_restrict = 0; %pocitadlo nesplneni apriornich omezene
 for ii=1:S_sim
    check_restrict = 0;
    while check_restrict==0
     pom = beta_0+norm_rnd(V_0);
     count_prior_restrict = count_prior_restrict+1;
     if prod(pom>=0)==1
        check_restrict=1;
        count_prior_restrict = count_prior_restrict-1;
     end
    end
 end
%integracni konstanta apriorni hustoty zohlednujici apriorni omezeni
int_prior_restrict = 1/(1-count_prior_restrict/...
    (S_sim+count_prior_restrict));
%korekce jmenovatele S-D pomeru hustot
 SDe_denom = int_prior_restrict*SDe_denom; 
  
%vypocet citatele na zaklade vystupu Gibbsova vzorkovace (doplneneho)
 E_SDe_nom = mean(SDe_nom);
%korekce citatele (neomezeneho normalniho rozdelenio) skrze
%integracni konstantu zohlednujici apriorni rozdeleni
%neni potreba ji pocitat znovu (spocitana v ramci Gibbsova vzorkovace)
 E_SDe_nom = 1/(1-count_post_restrict/(S+count_post_restrict-1))...
     *E_SDe_nom;
%Bayesuv faktor
 SD_e = E_SDe_nom/SDe_denom;
fprintf('\n Bayesuv faktor pro beta_reds = beta_trains \n')
fprintf('BF = %6.4f\n',SD_e)

%% Vypocet intervalu nejvyssi posteriorni hustoty
HPDI_90 = quantile([beta' h'],[0.05 0.95]);
fprintf('\n 90 procentni HPDI \n')
par_names = {'beta_0' 'beta_depart' 'beta_reds' 'beta_trains' 'h'};
for ii=1:length(par_names)
 fprintf('%15s %6.4f %6.4f\n',par_names{ii},HPDI_90(:,ii))   
end

File: 02_NLRM_s_nezavislou_normalni-gamma_apriorni_hustotou_Gibss/commute_testy2.py
# Nastaveni cesty do domovske slozky
import sys
root_dir = "/".join(sys.argv[0].split("/")[:-2])
sys.path.insert(1, root_dir + "/support")

# importy pomocnych funkci
from norm_rnd import norm_rnd
from gamm_rnd_koop import gamm_rnd_koop
from progress_info import progress_bar
from geweke import geweke

# importy oficialnich knihoven
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from tabulate import tabulate
from scipy.io import loadmat
from scipy.stats import norm
from scipy.stats import multivariate_normal as mvn
from numpy import transpose as t
from numpy.linalg import inv
from math import sqrt
from numpy import diag,\
                  ones,\
                  zeros

##### POZNAMKY:
# Predelat nacitani .mat souboru na CSV 


# Nacteni dat:
# time   ... cas cesty
# depart ... cas odchodu (v minutach od 6:30)
# reds   ... pocet cerevenych svetel na semaforu
# trains ... pocet vlaku, ktere musi nechat projek na Murrumbeena prejezdu

# Pro spravne nacteni datoveho souboru je treba spoustet script ze slozky 02_NLRM_s_nez...
# (popr. nastavit vlastni cestu)
raw_data = loadmat("data_commute.mat")

# Data jsou dictionary, kde hodnota je list hodnot
data = {}
for timeseries in ["reds", "time", "trains", "depart"]:
    data[timeseries] = raw_data[timeseries].transpose().tolist()[0]

# data jsou v matici v poradi: "reds", "time", "trains", "depart"
# data = np.array(data).transpose()

# ===== 1. Nastaveni apriornich hyperparametru a Gibbsova vzorkovace =====
# p(beta) ~ N(beta_0, v_0)
# p(h) ~ G(h_0, nu_0)
beta_0 = np.array([[30], [1], [3], [5]])
v_0 = diag([7.5**2, 0.25**2, 1**2, 2**2])
nu_0 = 40
s2_0 = 10**2
h_0 = 1 / s2_0

# Definice modelu
# Data musi byt zadefinovana jako matice i pokud se jedna o vektor
y = np.array([data["time"]]).transpose()
x = np.array([
        [1]*len(y),
        data["depart"],
        data["reds"],
        data["trains"]
    ]).transpose()

# Nastaveni Gibbsova vzorkovace
s = 50_000 + 1      # celkovy pocet generovanych vzorku + pocatecni hodnota
s_0 = 30_000 + 1    # pocet vyhozenych vzorku
s_1 = s - s_0       # pocet ponechanych vzorku

beta = np.array([[],[],[],[]])
beta = np.append(beta, beta_0, axis=1)

h = [h_0]

# Dalsi deklarace vektoru (pro Savage-Dickey pomer hustot)
sdd_nom = [0]                # jmenovatel pro SD pomer hustot (prvni prvek je nula, aby se predeslo chybe s indexovanim)
sde_nom = [0]                # citatel pro SD pomer hustot
count_post_restrict = 0      # pocitadlo nesplneni podminek apriornich restrikci na parametry

# ===== 2. Gibbsuv vzorkovac =====

print("Gibbsuv vzorkovac s apriorni restrikci:")
nu_1 = len(y) + nu_0    # (4.9) dle Koop (2003)
for i in range(1,s):
    # 1. blok Gibbsova vzorkovace
    # podminena hustota p(beta|h,y) ~ N(beta_1, v_1)
    v_1 = inv(inv(v_0) + h[i-1] * (t(x) @ x))                   # (4.4) dle Koop (2003)
    beta_1 = v_1 @ (inv(v_0) @ beta_0 + h[i-1] * (t(x) @ y))    # (4.5)

    check_restrict = True
    while check_restrict:
        new_beta = beta_1 + norm_rnd(v_1)                       # (4.7)
        count_post_restrict += 1
        if np.prod(new_beta >= 0):
            check_restrict = False
            count_post_restrict -= 1

    beta = np.append(beta, new_beta, axis=1)

    # 2. blok Gibbsova vzorkovace
    # Podminena hustota p(h|beta,y) ~ G(h_1, nu_1)
    h_1 = inv((1 / nu_1 * (t(y - x @ beta[:,[i]]) @ (y - x @ beta[:,[i]]) + nu_0 * 1 / h_0)))[0][0]     # (4.10)
    new_h = gamm_rnd_koop(h_1, nu_1, 1)[0]    # (4.8)   (funkce gamm_rnd_koop vraci array - je treba rozbalit)
    h.append(new_h)

    # Cast pro vypocet citatele SD pomeru hustot
    # d) beta_depart = 0 (druhy prvek vektoru beta)
    # p(b|h,y) ~ N(beta_1, v_1)
    v_1 = inv(inv(v_0) + h[i] * (t(x) @ x))                   # (4.4) dle Koop (2003)
    beta_1 = v_1 @ (inv(v_0) @ beta_0 + h[i] * (t(x) @ y))
    new_sdd_nom = norm.pdf(0, beta_1[1][0], sqrt(v_1[1][1]))
    sdd_nom.append(new_sdd_nom)

    # e) beta_reds = beta_trains
    # p(R*b|h,y) ~ N(R * beta_1, R * v_1 * R')
    r = np.array([[0, 0, 1, -1]])
    new_sde_nom = mvn.pdf(0, r @ beta_1, r @ v_1 @ t(r))
    sde_nom.append(new_sde_nom)

    progress_bar(i, s)

# ===== 3. Posteriorni analyza =====

# vyhozeni prvnich s_0 vzorku
beta = np.delete(beta, range(s_0), axis=1)
h = h[s_0:]
sdd_nom = sdd_nom[s_0:]
sde_nom = sde_nom[s_0:]

# Graficke zobrazeni konvergence
k = 100     # delka kroku (neplest si s krokem ve funkci geweke)
fig_1 = plt.figure()
axs = fig_1.subplots(nrows=3, ncols=2)

for i in range(beta.shape[0]):
    row = 0 if i < 2 else 1
    col = i%2
    axs[row, col].plot(range(0, beta.shape[1], k), beta[i,::k])
    axs[row, col].set_title(f"beta_{i+1}")

axs[2,0].plot(range(0, len(h), k), h[::k])
axs[2,0].set_title("h")

axs[2,1].remove()
fig_1.tight_layout()

# Gewekova konvergencni diagnostika
cd_beta = geweke(beta)
cd_h = geweke(np.array([h]))

# ===== 4. Prezentace vysledku =====
# apriorni str. hodnoty a sm. odchylky
# beta_0, h_0 - apriorni stredni hodnoty
std_beta_0 = [sqrt(element) for element in np.diag(v_0)]
std_h_0 = sqrt(2 * h_0**2 / nu_0)

# posteriorni str. hodnoty a sm. odchylky
mean_beta_1 = np.mean(beta, axis=1)
mean_h_1 = np.mean(h)

std_beta_1 = np.std(beta, axis=1)
std_h_1 = np.std(h)

headers = ["Parametr", "Prior. mean", "Prior. std", "Post. mean", "Post. std", "CD"]
table = []

for i in range(beta_0.shape[0]):
    table.append(
        [
            f"beta_{i+1}",
            round(beta_0[i][0], 4),
            round(std_beta_0[i], 4),
            round(mean_beta_1[i], 4),
            round(std_beta_1[i], 4),
            round(cd_beta[i]["cd"], 4)
        ]
    )

table.append(
    [
        "h",
        round(h_0, 4),
        round(std_h_0, 4),
        round(mean_h_1, 4),
        round(std_h_1, 4),
        round(cd_h["cd"], 4)
    ]
)
print("\nVysledne statistiky:")
print(tabulate(table, headers, tablefmt="pretty"))

# ===== 5. Test hypotezy ze beta_reds (beta_3) >= 2 =====
# pro ilustraci: zobrazeni rozdeleni beta_3 pomoci histogramu

fig_2 = plt.figure()
fig_2.suptitle("Posteriorni hustota parametru beta_3 (reds)")
axh = fig_2.add_subplot(1,1,1)
axh.hist(beta[2], bins=100, edgecolor="black", linewidth=1)
axh.set_xlabel("beta_3")

# pravdepodobnost beta_3 >= 2
pc = sum(beta[2] >= 2) / s_1
print("Pravdepodobnost beta_red >= 2 a odpovidajici Bayesuv faktor:")
print(f"- Pravdepodobnost = {round(pc, 4)}\n- Bayesuv faktor = {round(pc / (1 - pc), 4)}\n")

# ===== 6. Test hypotezy beta_depart (beta_2) == 0  =====
# Savage-Dickey pomer hustot
# Vyhodnotime p(beta_depart=0|M2)
# Apriorni hustota p(beta_depart)~N(beta_0(2),V_0(2,2))
sdd_denom = norm.pdf(0, beta_0[1][0], sqrt(v_0[1][1]))
s_sim = 50_000
count_prior_restrict = 0    # pocitadlo nesplneni apriornich omezeni

print("Simulace pro test hypotezy beta_depart (beta_2) == 0:")
for i in range(s_sim):
    check_restrict = True
    while check_restrict:
        pom = beta_0[1][0] + norm_rnd(np.array([[v_0[1][1]]]))
        count_prior_restrict += 1
        if pom >= 0:
            check_restrict = False
            count_prior_restrict -= 1
    progress_bar(i, s_sim)

# Integracni konstanta apriorni hustoty zohlednujici apriorni omezeni
int_prior_restrict = 1 / (1 - count_prior_restrict / (s_sim + count_prior_restrict))
# Korekce jmenovatele S-D pomeru hustot
sdd_denom = int_prior_restrict * sdd_denom
# Vypocet citatele na zaklade vystupu Gibbsova vzorkovace (rozsireneho)
e_sdd_nom = np.mean(sdd_nom)
# Integracni konstanta posteriorni hustoty zohlednujici apriorni omezeni
e_sdd_nom = 1 / (1 - count_post_restrict / (s + count_post_restrict - 1)) * e_sdd_nom
# Bayesuv faktor
sd_d = e_sdd_nom / sdd_denom

print("\nBayesuv faktor pro beta_depart = 0:")
print(f" - BF = {round(sd_d, 4)}")

# ===== 7. Test beta_reds (beta_3) == beta_trains (beta_4)  =====
# Linearni omezeni R*beta = r
# matice omezeni (na cely vektor parametru beta)
r = np.array([[0, 0, 1, -1]])
sde_denom = mvn.pdf(0, r @ beta_0, r @ v_0 @ t(r))
s_sim = 50_000
count_prior_restrict = 0

print("\nSimulace pro test beta_reds == beta_trains:")
for i in range(s_sim):
    check_restrict = True
    while check_restrict:
        pom = beta_0 + norm_rnd(v_0)
        count_prior_restrict += 1
        if np.prod(pom >= 0) == 1:
            check_restrict = False
            count_prior_restrict -= 1
    progress_bar(i, s_sim)

int_prior_restrict = 1 / (1 - count_prior_restrict / (s_sim + count_prior_restrict))
sde_denom = int_prior_restrict * sde_denom
e_sde_nom = np.mean(sde_nom)
e_sde_nom = 1 / (1 - count_post_restrict / (s + count_post_restrict - 1)) * e_sde_nom
sd_e = e_sde_nom / sde_denom

print("Bayesuv faktor pro beta_reds = beta_trains:")
print(f"- BF = {round(sd_e, 4)}")

# ===== 8. Vypocet intervalu nejvyssi posteriorni hodnoty =====

# vypocet kvantilu
hpdi_90_coeff = []
for coefficient in beta:
    hpdi_90_coeff.append(np.quantile(coefficient, [0.05,0.95]))

hpdi_90_h = np.quantile(h, [0.05,0.95])

headers = ["Parametr", "5% kvantil", "95% kvantil"]
table = []
for i in range(len(hpdi_90_coeff)):
    table.append([f"beta_{i+1}", round(hpdi_90_coeff[i][0], 4), round(hpdi_90_coeff[i][1], 4)])

table.append([f"h", round(hpdi_90_h[0], 4), round(hpdi_90_h[1], 4)])
print("\nTabulka 90% kvantilu pro odhadnute paramatry:")
print(tabulate(table, headers, tablefmt="pretty"))

# Vykresleni grafu
plt.show()
